loopy/transform/iname.py

```diff
@@ -1172,42 +1197,45 @@ def rename_iname(kernel, old_iname, new_iname, existing_ok=False, within=None):
 
         # }}}
 
-        from pymbolic import var
-        subst_dict = {old_iname: var(new_iname)}
-
-        from loopy.match import parse_stack_match
-        within = parse_stack_match(within)
+    from pymbolic import var
+    subst_dict = {old_iname: var(new_iname) for old_iname in old_inames}
 
-        from pymbolic.mapper.substitutor import make_subst_func
-        rule_mapping_context = SubstitutionRuleMappingContext(
-                kernel.substitutions, var_name_gen)
-        smap = RuleAwareSubstitutionMapper(rule_mapping_context,
-                        make_subst_func(subst_dict), within)
+    from loopy.match import parse_stack_match
+    within = parse_stack_match(within)
 
-        kernel = rule_mapping_context.finish_kernel(
-                smap.map_kernel(kernel))
+    from pymbolic.mapper.substitutor import make_subst_func
+    rule_mapping_context = SubstitutionRuleMappingContext(
+            kernel.substitutions, var_name_gen)
+    smap = RuleAwareSubstitutionMapper(rule_mapping_context,
+                    make_subst_func(subst_dict), within)
 
-        new_instructions = []
-        for insn in kernel.instructions:
-            if (old_iname in insn.within_inames
-                    and within(kernel, insn, ())):
-                insn = insn.copy(
-                        within_inames=(
-                            (insn.within_inames - frozenset([old_iname]))
-                            | frozenset([new_iname])))
+    from loopy.kernel.instruction import MultiAssignmentBase
 
-            new_instructions.append(insn)
+    def check_insn_has_iname(kernel, insn, *args):
+        return (not isinstance(insn, MultiAssignmentBase)
+                or len(frozenset(old_inames) & insn.dependency_names()) != 0
+                or len(frozenset(old_inames) & insn.reduction_inames()) != 0)
 
-        kernel = kernel.copy(instructions=new_instructions)
+    kernel = rule_mapping_context.finish_kernel(
+            smap.map_kernel(kernel, within=check_insn_has_iname))
 
-    else:
-        kernel = duplicate_inames(
-                kernel, [old_iname], within=within, new_inames=[new_iname])
+    new_instructions = [insn.copy(within_inames=((insn.within_inames
+                                                  - frozenset(old_inames))
+                                                 | frozenset([new_iname])))
+                        if ((len(frozenset(old_inames) & insn.within_inames) != 0)
```

Comment:
* Repeated casts to `frozenset`?
* Why the odd `len(...) != 0` construction (instead of just `...`)?

Reply:
> Repeated casts to frozenset?

old_inames is stored as `list`, didn't want to use another variable specifically for the cast. (Kept as is)

> Why the odd len(...) != 0 construction (instead of just ...)?

Thought the former was bit more intuitive. However, changed to the latter.